# Atividade Kera (Patrick)

In [9]:
!pip install pandas tensorflow numpy 

Defaulting to user installation because normal site-packages is not writeable


In [10]:
# Importação de bibliotecas
import pandas as pd
import tensorflow as tf
from tensorflow import keras
from keras import layers
from keras import metrics
import numpy as np

## Importação dos dados

In [11]:
df = pd.read_csv('mushroom_cleaned.csv')

Visualização dos dados:

In [12]:
df

cap-diameter  cap-shape  gill-attachment  gill-color  stem-height  \
0              1372          2                2          10     3.807467   
1              1461          2                2          10     3.807467   
2              1371          2                2          10     3.612496   
3              1261          6                2          10     3.787572   
4              1305          6                2          10     3.711971   
...             ...        ...              ...         ...          ...   
54030            73          5                3           2     0.887740   
54031            82          2                3           2     1.186164   
54032            82          5                3           2     0.915593   
54033            79          2                3           2     1.034963   
54034            72          5                3           2     1.158311   

       stem-width  stem-color    season  class  
0            1545          11  1.804273      1  
1            1557          11  1.804273      1  
2            1566          11  1.804273      1  
3            1566          11  1.804273      1  
4            1464          11  0.943195      1  
...           ...         ...       ...    ...  
54030         569          12  0.943195      1  
54031         490          12  0.943195      1  
54032         584          12  0.888450      1  
54033         491          12  0.888450      1  
54034         492          12  0.888450      1  

[54035 rows x 9 columns]

## Explicando o conjunto de dados

O conjunto de dados escolhido possui dados sobre 54035 cogumelos tendo assim 534035 linhas de dados. Os quais cada uma possui 9 colunas as quais representam características do cogumelo como:


* Cap Diameter: Diâmetro do chapéu do cogumelo.
* Cap Shape: Forma do chapéu do cogumelo.
* Gill Attachment: Como as lamelas estão ligadas ao caule.
* Gill Color: Cor das lamelas do cogumelo.
* Stem Height: Altura do caule do cogumelo.
* Stem Width: Largura do caule do cogumelo.
* Stem Color: Cor do caule do cogumelo.
* Season: Época do ano em que o cogumelo aparece.


As quais são utilizadas visando a identificação do target "class", o qual classifica o cogumelo como comestível ou venenoso, sendo 0 para comestível e 1 para venenoso.


## Criação do modelo

Abaixo está sendo criado o modelo com as métricas e componentes solicitados, sendo eles:


* camada dense(1):


Camada Dense: Camada totalmente conectada onde todos os neurônios têm ligações com a camada anterior, contribuindo para a captura de características dos dados. Neste caso, a camada possui uma única unidade de saída.


* função de ativação(sigmoid):


Função utilizada para calcular a saída do modelo. No caso da sigmoid, esse valor é mapeado para o intervalo entre 0 e 1.


* Otimizador (adam):


Ajusta dinamicamente as taxas de aprendizado para cada peso, otimizando a convergência e melhorando o treinamento do modelo.


* função de Perda (binary_crossentropy):


Mede a diferença entre as previsões do modelo e os valores reais, sendo adequada para problemas de classificação binária.


* métrica (accuracy):


Métrica utilizada para avaliar o modelo, no caso, a acurácia, que é baseada na porcentagem de previsões corretas feitas pelo modelo.


* métrica (F1):


Métrica utilizada para avaliação do modelo. A F1 combina precisão e recall, fornecendo uma métrica que leva em consideração ambos.

In [13]:
#Criando o modelo com 1 camada dense e função de ativação sigmoid
model = keras.Sequential([
    layers.Dense(units=1, activation='sigmoid')
])

# Compilando o modelo com otimizador adam, função de perda binary_crossentropy e metricas accuracy e f1
model.compile(optimizer='adam',
              loss='binary_crossentropy',
              metrics=['accuracy', metrics.F1Score()])

## Treinamento de modelo

Primeiramente para evitar qualquer enviesamento do modelo os conjuntos de treino e teste serão separados, sendo 20% para teste e 80% para treino:


In [14]:
# seprando variavel alvo
X = df.drop('class', axis=1).values
y = df['class'].values
y = np.expand_dims(y, axis=-1)

# Dividindo os em treino e teste
dataset_size = len(X)
train_size = int(dataset_size * 0.8)

X_train, X_test = X[:train_size], X[train_size:]
y_train, y_test = y[:train_size], y[train_size:]

A seguir será realizado o treinamento do modelo, nas condições solicitadas 50 épocas e 10 de batch size:

In [15]:
# Realizando treinamento do modelo, com base nas condições fornecidas
model.fit(X_train, y_train, epochs=50, batch_size=10, validation_data=(X_test, y_test))

Epoch 1/50
4323/4323 ━━━━━━━━━━━━━━━━━━━━ 3s 583us/step - accuracy: 0.5472 - f1_score: 0.7156 - loss: 15.1403 - val_accuracy: 0.6790 - val_f1_score: 0.6656 - val_loss: 1.1150
Epoch 2/50
4323/4323 ━━━━━━━━━━━━━━━━━━━━ 2s 561us/step - accuracy: 0.5891 - f1_score: 0.7186 - loss: 0.7129 - val_accuracy: 0.7380 - val_f1_score: 0.6656 - val_loss: 0.7091
Epoch 3/50
4323/4323 ━━━━━━━━━━━━━━━━━━━━ 2s 562us/step - accuracy: 0.5949 - f1_score: 0.7236 - loss: 0.7034 - val_accuracy: 0.5026 - val_f1_score: 0.6656 - val_loss: 0.6819
Epoch 4/50
4323/4323 ━━━━━━━━━━━━━━━━━━━━ 2s 563us/step - accuracy: 0.5965 - f1_score: 0.7181 - loss: 0.7116 - val_accuracy: 0.7196 - val_f1_score: 0.6656 - val_loss: 0.6821
Epoch 5/50
4323/4323 ━━━━━━━━━━━━━━━━━━━━ 2s 556us/step - accuracy: 0.6011 - f1_score: 0.7220 - loss: 0.6986 - val_accuracy: 0.4974 - val_f1_score: 0.6656 - val_loss: 0.7496
Epoch 6/50
4323/4323 ━━━━━━━━━━━━━━━━━━━━ 2s 563us/step - accuracy: 0.5993 - f1_score: 0.7185 - loss: 0.6975 - val_accuracy: 0.73

Teste do modelo treinado, fornecendo tanto acurácia quanto o F1 Score:

In [16]:
# Utilização e avaliação do modelo treinado com base no conjunto de teste
results = model.evaluate(X_test, y_test)

338/338 ━━━━━━━━━━━━━━━━━━━━ 0s 423us/step - accuracy: 0.7012 - f1_score: 0.4774 - loss: 0.7585 


## Análise dos resultados e possíveis melhorias

O modelo alcançou uma acurácia de cerca de 0.70 e um F1 Score de aproximadamente 0.48. Isso significa que o modelo acerta em 70% dos casos ao identificar se um cogumelo é venenoso ou não. E o F1 Score de 0,48 indica que o modelo está alcançando um certo equilíbrio entre precisão e capacidade de detectar corretamente os casos positivos.

Na minha visão, o modelo obteve resultados razoáveis considerando sua simplicidade , mas ainda há grande espaço para melhoras.


Sendo algumas possíveis melhorias que poderiam ser implementadas:


* Avaliação do modelo através do recall, uma vez que essa métrica da grande pesos aos falsos negativos, que no contexto da atividade são os mais nocivos uma vez que deixa cogumelos venenosos passarem despercebidos.

* Teste de outras funções de ativação, para ver como elas impactam o modelo.

* Teste de outros otimizadores.

* Aplicação de tratamento nos dados como one hot encoding e normalização.

* E principalmente a adição de mais camadas no modelo, para um melhor aproveitamento da diversidade de características fornecidas.
